src/lib/yacc/grammar.rs

```diff
@@ -421,6 +432,171 @@ impl Grammar for YaccGrammar {
     }
 }
 
+/// A `SentenceGenerator` can generate minimal sentences for any given non-terminal. e.g. for the
+/// grammar:
+///
+/// ```text
+/// %start A
+/// %%
+/// A: A B | ;
+/// B: C | D;
+/// C: 'x' B | 'x';
+/// D: 'y' B | 'y' 'z';
+/// ```
+///
+/// the following are valid minimal sentences:
+///
+/// ```text
+/// A: []
+/// B: [x]
+/// C: [x]
+/// D: [y, x] or [y, z]
+/// ```
+pub struct SentenceGenerator<'a> {
+    grm: &'a YaccGrammar,
+    nonterm_costs: Vec<u64>,
+    term_costs: Vec<u64>
+}
+
+impl<'a> SentenceGenerator<'a> {
+    fn new<F>(grm: &YaccGrammar, term_cost: F) -> SentenceGenerator
+        where F: Fn(TIdx) -> u64
+    {
+        let nt_costs = nonterm_costs(grm, &term_cost);
+        let mut term_costs = Vec::with_capacity(grm.terms_len());
+        for i in 0..grm.terms_len() {
+            term_costs.push(term_cost(TIdx::from(i)));
+        }
+        SentenceGenerator{grm, term_costs, nonterm_costs: nt_costs}
```

Comment:
Why are you not initialising this with `SentenceGenerator{grm, nt_costs, term_costs}`?

Reply:
It's the way Rust struct initialisation works (in recent versions of rustc): if a field and a variable name *happen* to share the name, you don't need to specify both; but in this case, they don't.

Reply:
Ah so the arguments are unordered? Meaning `SentenceGenerator{grm, nt_costs, term_costs}` won't work?

Reply:
Exactly.